### DISCUSSION TITLE
Inconsistent cascading delete behavior for folders in HTTP API

### DISCUSSION THREAD

**Discussion by jacobstr** [grafana/grafana#64660] [General]
* Deleting a folder with alert rules defined will result in a kind of cascading deletion protection (see snippet below).
* However, deleting a folder with only dashboards defined does not offer this same protection. The dashboards within will happily be deleted.

E.g. when deleting a folder with an alert, terraform ([grafana_folder](https://registry.terraform.io/providers/grafana/grafana/latest/docs/resources/folder)) (which ultimately calls into these APIs) responds with:

```
status: 400, body: {"message":"folder cannot be deleted: folder contains alert rules"}
```

Note: originally I was feeling like submitting this in the `grafana-terraform-provider`, but this level of validation felt more appropriate closer to the HTTP backend API.

Note: I realize this is not a straightforward backwards-compatible change to make. I hear there's some nested folder work ongoing and would be curious if that refactors affects any of these behaviors.

TLDR; There's a bit of a violation in principle of least surprise here. It's odd that alert rules "block" folder deletes, but dashboards don't. 

---

Note: This isn't the only case of differing behavior for folders when they're used for alerting rules vs. dashboards - there's also an odd restriction where alerts can't be placed into folders that contain `/` in the name. Overall, it makes me _suspect_ that the folder model has evolved to become somewhat [stovepipe](https://en.wikipedia.org/wiki/Stovepipe_system) rather than a kind of homogenous box that behaves similarly regardless of whether it contains dashboards, other folders (upcoming?), alerts, or any new entity type that might get dreamed up in the future.

Pardon the pejorative use of "stovepipe" - just trying to convey the UX-impression imparted by these differing behaviors 😅.

**Comment by zsofiakm** [comment]
Hi @jacobstr,
Thanks for your feedback! I agree that these behaviors are inconsistent. We're looking at unifying delete permissions, and this will be addressed as part of the work on subfolders, which we expect to make generally available during this fall.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82658. Feel free to continue the discussion around this there. Thank you!
